Index: linux-2.6.29/sound/soc/davinci/davinci-i2s.c
===================================================================
--- linux-2.6.29.orig/sound/soc/davinci/davinci-i2s.c	2009-12-10 10:01:05.110332216 -0600
+++ linux-2.6.29/sound/soc/davinci/davinci-i2s.c	2009-12-10 10:01:48.492918693 -0600
@@ -104,10 +104,12 @@
 
 static struct davinci_pcm_dma_params davinci_i2s_pcm_out = {
 	.name = "I2S PCM Stereo out",
+	.data_type = 2,
 };
 
 static struct davinci_pcm_dma_params davinci_i2s_pcm_in = {
 	.name = "I2S PCM Stereo in",
+	.data_type = 2,
 };
 
 struct davinci_mcbsp_dev {
