Index: linux-2.6.29/drivers/usb/musb/davinci.c
===================================================================
--- linux-2.6.29.orig/drivers/usb/musb/davinci.c	2009-11-02 08:04:04.916614979 -0600
+++ linux-2.6.29/drivers/usb/musb/davinci.c	2009-11-02 08:11:14.132412584 -0600
@@ -407,6 +407,13 @@
 	musb->board_set_vbus = davinci_set_vbus;
 	davinci_source_power(musb, 0, 1);
 
+	/* dm355 Leopard support for usb host */
+	if(machine_is_dm355_leopard()){
+		 u32	phy_ctrl = __raw_readl(USB_PHY_CTRL);
+		phy_ctrl &= ~(3 << 9);
+		__raw_writel(phy_ctrl, USB_PHY_CTRL);
+	}
+
 	/* dm355 EVM swaps D+/D- for signal integrity, and
 	 * is clocked from the main 24 MHz crystal.
 	 */
